recommenders/models/sasrec/util.py

```diff
@@ -0,0 +1,309 @@
+        seq, padding="pre", truncating="pre", maxlen=seq_max_len
+    )
+    pos = tf.keras.preprocessing.sequence.pad_sequences(
+        pos, padding="pre", truncating="pre", maxlen=seq_max_len
+    )
+    neg = tf.keras.preprocessing.sequence.pad_sequences(
+        neg, padding="pre", truncating="pre", maxlen=seq_max_len
+    )
+
+    inputs["users"] = np.expand_dims(np.array(u), axis=-1)
+    inputs["input_seq"] = seq
+    inputs["positive"] = pos
+    inputs["negative"] = neg
+
+    target = np.concatenate(
+        [
+            np.repeat(1, seq.shape[0] * seq.shape[1]),
+            np.repeat(0, seq.shape[0] * seq.shape[1]),
+        ],
+        axis=0,
+    )
+    target = np.expand_dims(target, axis=-1)
+    return inputs, target
+
+
+def train(model, dataset, sampler, **kwargs):
+    """
+    High level function for model training
+    """
+    num_epochs = kwargs.get("num_epochs", 10)
+    maxlen = kwargs.get("maxlen", 100)
+    num_neg_test = kwargs.get("num_neg_test", 100)
+    batch_size = kwargs.get("batch_size", 128)
+    lr = kwargs.get("learning_rate", 0.001)
+
+    [user_train, user_valid, user_test, usernum, itemnum] = dataset
+    num_steps = int(len(user_train) / batch_size)
+
+    optimizer = tf.keras.optimizers.Adam(
+        learning_rate=lr, beta_1=0.9, beta_2=0.999, epsilon=1e-7
+    )
+
+    loss_function = model.loss_function
+
+    train_loss = tf.keras.metrics.Mean(name="train_loss")
+
+    train_step_signature = [
+        {
+            "users": tf.TensorSpec(shape=(None, 1), dtype=tf.int64),
+            "input_seq": tf.TensorSpec(shape=(None, maxlen), dtype=tf.int64),
+            "positive": tf.TensorSpec(shape=(None, maxlen), dtype=tf.int64),
+            "negative": tf.TensorSpec(shape=(None, maxlen), dtype=tf.int64),
+        },
+        tf.TensorSpec(shape=(None, 1), dtype=tf.int64),
+    ]
+
+    @tf.function(input_signature=train_step_signature)
+    def train_step(inp, tar):
+        with tf.GradientTape() as tape:
+            pos_logits, neg_logits, loss_mask = model(inp, training=True)
+            loss = loss_function(pos_logits, neg_logits, loss_mask)
+
+        gradients = tape.gradient(loss, model.trainable_variables)
+        optimizer.apply_gradients(zip(gradients, model.trainable_variables))
+
+        train_loss(loss)
+        return loss
+
+    T = 0.0
+    t0 = time.time()
+
+    for epoch in range(1, num_epochs + 1):
+
+        step_loss = []
+        train_loss.reset_states()
+        for step in tqdm(
+            range(num_steps), total=num_steps, ncols=70, leave=False, unit="b"
+        ):
+
+            u, seq, pos, neg = sampler.next_batch()
+
+            inputs, target = create_combined_dataset(u, seq, pos, neg, maxlen)
+
+            loss = train_step(inputs, target)
+            step_loss.append(loss)
+
+        #     print(f"Epoch: {epoch}, Loss: {np.mean(step_loss):.3f}, {train_loss.result():.3f}")
+
+        if epoch % 2 == 0:
+            t1 = time.time() - t0
```

Comment:
minor thing, we have a Timer class: https://github.com/microsoft/recommenders/blob/main/recommenders/utils/timer.py